db-engine/src/main/java/com/splicemachine/db/iapi/sql/dictionary/IndexRowGenerator.java

```diff
@@ -230,49 +271,69 @@ private void getIndexRowHelper(ExecRow baseRow,
 							boolean alwaysIncludeLocation)
 						throws StandardException
 	{
-		/*
-		** Set the columns in the index row that are based on columns in
-		** the base row.
-		*/
-		int[] baseColumnPositions = id.baseColumnPositions();
-		int colCount = baseColumnPositions.length;
-
-		if (bitSet == null)
-		{
-			/*
-			** Set the columns in the index row that are based on columns in
-			** the base row.
-			*/
-			for (int i = 0; i < colCount ; i++)
-			{
-				indexRow.setColumn(i + 1,
-						baseRow.getColumn(baseColumnPositions[i]));
+		int colCount;
+		if (isOnExpression()) {
+			colCount = isAscending().length;
+			ExecRow expandedRow;
+
+			if (bitSet == null) {
```

Comment:
maybe extract function getExpandedRow() ?

Reply:
why is this code so different from the code in lines 326-336? it somehow has the feel it's doing the same, it should maybe look the same, or maybe use same code (e.g. extract a function which returns an array of pairs expandedRowIndex -> column, so that i could write e.g. code like

```
expandedRow = new ValueRow(maxNumCols);
auto mapping = getRowColMapping();
for( auto pair : mapping )
{
  expandedRow.setColumn(pair.first, pair.second);
}
```

and the code below would be

```
auto mapping = getRowColMapping();
for( auto pair : mapping )
{
  indexRow.setColumn(pair.first, pair.second);
}
```

Reply:
The code in lines 326 - 336 (in latest commit 329 - 339) doesn't actually expand `baseRow`. Consider a partial `baseRow` that has the original columns 3, 5, 7. The row looks like this:

```
original col pos    :  3   |   5   |   7
row column pos   :  0   |   1    |   2
```
`bitSet` should have bits for 3, 5, and 7 set. Let's suppose the index column to base column mapping (i.e., `baseColumnPositions`) is 5, 3, 7. That is, the first index column is full base row column 5, the second index column is full base row column 3, and so on. Btw, column position is always 1-based.

The code tries to assign columns in `indexRow` directly from the partial row. To achieve that, it needs to get the position of the full base row column in this partial row. Take the first index column as an example:

`fullColumnNumber = baseColumnPositions[0]`, which is 5. Now we count the bitSet from 1 to 5 inclusive. 3 is set, 5 is set, so we get `partialColumnNumber = 2`. That is, the full base row column 5 is at position 2 in this partial row.

The new code doesn't do this. Instead, it creates a new row, trying to be wide enough to hold the columns at their full base row positions. Example of an expanded row:

```
original column pos :  x | x | x | 3 | x | 5 | x | 7
expanded row pos   :  0 | 1 | 2 | 3 | 4 | 5 | 6 | 7
```

We need to expand the row as a way to normalize the input for generated code. During code generation, we only know full base row column position, not with any mapping. So for generated code, it assumes the row passed in is a full base row.

We can unify the two code paths by always expanding the row. But then it would be unnecessary extra cost in case of normal index creation. That's one point I mentioned in the design discussion the other day.

Reply:
ah, ok